File: solutions/ruby/house/1/house.rb
class House
  FULL = [
    {action: "",                 title: "the horse and the hound and the horn"},
    {action: "that belonged to", title: "the farmer sowing his corn"},
    {action: "that kept",        title: "the rooster that crowed in the morn"},
    {action: "that woke",        title: "the priest all shaven and shorn"},
    {action: "that married",     title: "the man all tattered and torn"},
    {action: "that kissed",      title: "the maiden all forlorn"},
    {action: "that milked",      title: "the cow with the crumpled horn"},
    {action: "that tossed",      title: "the dog"},
    {action: "that worried",     title: "the cat"},
    {action: "that killed",      title: "the rat"},
    {action: "that ate",         title: "the malt"},
    {action: "that lay in",      title: "the house that Jack built."}
  ]
  def self.recite
    12.times.map { |i| verse(i) }.join("\n")
  end

  def self.verse(index)
    verses = FULL.drop(FULL.size - index - 1)
    first = verses[0]
    rest = verses.drop(1)

    "This is #{first[:title]}\n" +
      (rest.empty? ? ""
       : rest.each_with_object([]) { |h, a| a << "#{h[:action]} #{h[:title]}" }
             .join("\n") + "\n"
      )
  end
end

File: solutions/zig/house/1/house.zig
const std = @import("std");

const verses: [12][]const u8 = .{
    \\This is the house that Jack built.
    ,
    \\This is the malt that lay in the house that Jack built.
    ,
    \\This is the rat that ate the malt that lay in the house that Jack built.
    ,
    \\This is the cat that killed the rat that ate the malt that lay in the house that Jack built.
    ,
    \\This is the dog that worried the cat that killed the rat that ate the malt that lay in the house that Jack built.
    ,
    \\This is the cow with the crumpled horn that tossed the dog that worried the cat that killed the rat that ate the malt that lay in the house that Jack built.
    ,
    \\This is the maiden all forlorn that milked the cow with the crumpled horn that tossed the dog that worried the cat that killed the rat that ate the malt that lay in the house that Jack built.
    ,
    \\This is the man all tattered and torn that kissed the maiden all forlorn that milked the cow with the crumpled horn that tossed the dog that worried the cat that killed the rat that ate the malt that lay in the house that Jack built.
    ,
    \\This is the priest all shaven and shorn that married the man all tattered and torn that kissed the maiden all forlorn that milked the cow with the crumpled horn that tossed the dog that worried the cat that killed the rat that ate the malt that lay in the house that Jack built.
    ,
    \\This is the rooster that crowed in the morn that woke the priest all shaven and shorn that married the man all tattered and torn that kissed the maiden all forlorn that milked the cow with the crumpled horn that tossed the dog that worried the cat that killed the rat that ate the malt that lay in the house that Jack built.
    ,
    \\This is the farmer sowing his corn that kept the rooster that crowed in the morn that woke the priest all shaven and shorn that married the man all tattered and torn that kissed the maiden all forlorn that milked the cow with the crumpled horn that tossed the dog that worried the cat that killed the rat that ate the malt that lay in the house that Jack built.
    ,
    \\This is the horse and the hound and the horn that belonged to the farmer sowing his corn that kept the rooster that crowed in the morn that woke the priest all shaven and shorn that married the man all tattered and torn that kissed the maiden all forlorn that milked the cow with the crumpled horn that tossed the dog that worried the cat that killed the rat that ate the malt that lay in the house that Jack built.
};

pub fn recite(buffer: []u8, start_verse: u32, end_verse: u32) []const u8 {
    var s = std.io.fixedBufferStream(buffer);
    var w = s.writer();
    for (verses[start_verse - 1 .. end_verse]) |verse| {
        if (s.pos > 0) w.print("\n", .{}) catch unreachable;
        w.print("{s}", .{verse}) catch unreachable;
    }
    return s.getWritten();
}
